File: 20230605_Monday/20230605.go
// Leetcode: 1232. Check If It Is a Straight Line.
// https://leetcode.com/problems/check-if-it-is-a-straight-line/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 7 ms, faster than 50.00% of Go online submissions for Check If It Is
// a Straight Line.
// Memory Usage: 3.6 MB, less than 34.21% of Go online submissions for Check If
// It Is a Straight Line.
// 2023.06.05 Daily Challenge.
package main

func gdx(a, b []int) int {
	return b[0] - a[0]
}
func gdy(a, b []int) int {
	return b[1] - a[1]
}
func checkStraightLine(coordinates [][]int) bool {
	c := coordinates
	dx := gdx(c[0], c[1])
	dy := gdy(c[0], c[1])
	for i := 2; i < len(c); i += 1 {
		one := dx * gdy(c[0], c[i])
		two := dy * gdx(c[0], c[i])
		if one != two {
			return false
		}
	}
	return true

}

File: 20230605_Monday/20230605.rb
# Leetcode: 1232. Check If It Is a Straight Line.
# https://leetcode.com/problems/check-if-it-is-a-straight-line/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 75 ms, faster than 75.00% of Ruby online submissions for Check If It
# Is a Straight Line.
# Memory Usage: 210.9 MB, less than 50.00% of Ruby online submissions for Check
# If It Is a Straight Line.
# 2023.06.05 Daily Challenge.
# @param {Integer[][]} coordinates
# @return {Boolean}
def gdx(a,b) # get dx;
  b[0] - a[0]
end
def gdy(a,b) # get dy;
  b[1] - a[1]
end

def check_straight_line(coordinates)
  # Based on:
  # https://leetcode.com/problems/check-if-it-is-a-straight-line/solution/
  c = coordinates
  dx = gdx(c[0], c[1])
  dy = gdy(c[0], c[1])
  (2...c.size).each do |i|
    if dx * gdy(c[0],c[i]) != dy * gdx(c[0],c[i])
      return false
    end
  end
  return true
end
